### ISSUE TITLE
MIT-MAGIC-COOKIE file location is unpredictable

### ISSUE THREAD

**Issue by meyergru** [sddm/sddm#622]
The xauthority (or MIT-MAGIC-COOKIE) file is no longer at /var/run/sddm/:0 but at an unpredictable location like /var/run/sddm/{ca7d6d45-9b4d-4a92-997e-c9c11081dea1} which is build by including a random UUID in the path.

This causes interoperability problems. When trying to start x11vnc, for example, that file cannot be found. E.g. with Ubuntu 15.10, one could start x11vnc with '/usr/bin/x11vnc -auth /var/run/sddm/:0 -display :0 -rfbauth /etc/vncpasswd -rfbport 5943 -reopen -forever', this is no longer possible with Ubuntu 16.04.

Downstream bug here:
https://bugs.launchpad.net/ubuntu/+source/sddm/+bug/1566542

**Comment by BETLOG** [comment]
Also experiencing this problem.

**Comment by VVD** [comment]
Same here.

**Comment by DejavuDmitry** [comment]
Same.... Please help!!! =)

**Comment by plfiorini** [comment]
$XAUTHORITY contains the Xauthority location

**Comment by meyergru** [comment]
That would be the case if a user was to start x11vnc - I addressed the case where this is done via a systemd script on the machine. In that case, you don't have $XAUTHORITY.

**Comment by plfiorini** [comment]
@meyergru You should use Xsession for that

**Comment by VVD** [comment]
One more place in config to support it manually… Every time new place… It's so fascinatingly to search every year how to configure auto start for x11vnc!

**Comment by plfiorini** [comment]
Xsession exists for ages

**Comment by meyergru** [comment]
IMHO: It is anyone's choice to have their software used or not.

To put it clear: Most standard distributions (Ubuntu, OpenSUSE) have x11vnc on board, but they do not cover/care how to get it working as it was intended, namely: to have an X11 server on a headless machine that can be reached via network from, for example, a Windows management workstation. 

In the past, it was relatively easy to get this going by providing a startup script that just starts x11vnc in the way I showed in my first posting. Now it is not. And since not everybody knows how to figure it out with Xsession, $XAUTHORITY or even Kerberos, the net result will be that x11vnc will be used less often. For me, lacking the time to do this over and over again for each new distribution release and/or init system, that will most certainly be the case.

Making software more useable, is like writing a good book: The effort to write it is worth the time, because it is written only once, but read probably thousands of times. So, spare your readers some time by investing it while writing.

In case you did not notice: I am talking about the fine line between bugs and lacking useability.

**Comment by plfiorini** [comment]
@meyergru Xauthority of the greeter was turned into a uuid with this commit 469ea132e847796a4f87e7f7a8f7e505c80f3071 probably related to c1d1c0c8050d45a6316217f2e2ada632b8eddb66 - since the display name cannot be decided earlier, SDDM lack this information and cannot name Xauthority like it used before.
We can't have ~sddm/.Xauthority either if we want to support multi-seat.

I can understand your frustration but if you are not talking about something that is not fully covered by distributions you might expect to adapt when things change. What I can do is try finding an alternative solution, not changing the greeter Xauthority name.

SDDM has a script called Xsetup and it is executed as root before the greeter window starts and the standard XAUTHORITY env var is passed. You can find the script here: /usr/share/sddm/scripts/Xsetup - try running x11vnc there instead.

**Comment by meyergru** [comment]
Neat trick. Not very general, since it works only for a certain display manager, but it works fine, once you insert this into /usr/share/sddm/scripts/Xsetup:

/usr/bin/x11vnc -rfbauth /etc/vncpasswd -rfbport 5943 -reopen -forever &

I can buy the multi-seat argument, BTW. Maybe I should accept x11vnc more as an addon to the starting of the X server than as a separate service under systemd.

**Comment by dpellegr** [comment]
Long time closed, but if someone googles this out like I did, here is a possible way to build a systemd service:

```
[Unit]
Description=x11vnc VNC Server for X11
Requires=display-manager.service
After=display-manager.service

[Service]
Type=forking
ExecStartPre=/usr/bin/bash -c "/usr/bin/systemctl set-environment SDDMXAUTH=$(/usr/bin/find /var/run/sddm/ -type f)"
ExecStart=/usr/bin/x11vnc -display :0 -auth ${SDDMXAUTH} -ncache 10 -forever -shared -bg -o /var/log/x11vnc.log -rfbauth /etc/x11vnc.passwd -xkb -norc -noxrecord -noxdamage -nomodtweak -multiptr
ExecStop=/usr/bin/killall x11vnc
Restart=on-failure
Restart-sec=2

[Install]
WantedBy=graphical.target

```

**Comment by starkline** [comment]
@dpellegr Thanks so much. This saved me from hours more frustration than I already spent. Much appreciated for sharing this solution.

**Comment by BETLOG** [comment]
@dpellegr I have recently been using x2x as a means of avoiding needing 2 mice and 2 keyboards when using my 2 computers with one screen above the other... but for login I still need both sets of keyboards. This has had me indirectly thinking back to this issue, so your post is a little bit of synchronicity.
I wonder if your solution could be leveraged into some way of letting me get rid of one keyboard entirely when using x2x?

**Comment by ramsamba** [comment]
@dpellegr  Thank you for your code, saved me a huge headache.

On implementing your systemd script, i kept getting the following error:
```
x11vnc.service: Start request repeated too quickly.
x11vnc.service: Failed with result 'exit-code'.
Failed to start x11vnc VNC Server for X11.
```

I fixed the problem based on this [article](https://medium.com/@benmorel/creating-a-linux-service-with-systemd-611b5c8b91d6) by setting the following:
`StartLimitIntervalSec=0`

I hope this post benefits others who face the same problem.

**Comment by LanThrusteR** [comment]
Same problem here. SDDM needs an event on re-creation of `/var/run/sddm/{magic-cookie}`
like Xsession then the problem with x11vnc can be solved properly 

When user logouts from KDE or LXQT or XFCE and SDDM greeting popups - x11vnc looses DISPLAY and you have to re-launch it manually from shell or you loose an ability to login via VNC viewer.

The best solution I could come up with RC is to 

add to `/etc/conf.d/x11vnc` an option -auth $(find /var/run/sddm/ -type f | head -n 1)

edit `/usr/share/sddm/scripts/Xsetup` and add
   (sleep 5; /etc/init.d/x11vnc zap; /etc/init.d/x11vnc restart) & 

This way you can get x11vnc running when you exit from session (like exit from KDE session)

From what I understand it could have been better with SDDM...,
to fix this problem we need to patch SDDM, it needs 

`# Path to a script to execute AFTER /var/run/sddm/{magic-cookie} is created 
MagicCookieCommand=/usr/share/sddm/scripts/Xcookie`

or alternatively SDDM needs to run Xsetup after /var/run/sddm/{magic-cookie} is created, not like it does now - before. 

That way we can fix x11vnc problem... or have a switch to get back to the previous behavior.... I can't understand how this could be overlooked in SDDM design. 

The problem is that currently you can't rely on anything, - X server presence, /var/run/sddm/{magic-cookie} presence, no, you can't check on them in Xsetup, you need just to wait and hope while greeting popups and the MagicCookie updated and then restart x11vnc. 

However you can lower 5 second pause if your SDDM is fast or get it higher in case it takes some time to boot (like with GMA cards)/var/run/sddm/{magic-cookie}

**Comment by LanThrusteR** [comment]
> SDDM has a script called Xsetup and it is executed as root before the greeter window starts and the standard XAUTHORITY env var is passed. You can find the script here: /usr/share/sddm/scripts/Xsetup - try running x11vnc there instead.

That was a poor advise.  It won't work.

**Comment by meyergru** [comment]
> That was a poor advise. It won't work.

Really? You are still using Ubuntu 16.04? That advice actually worked more than 6 years ago...

**Comment by LanThrusteR** [comment]
> > That was a poor advise. It won't work.
> 
> Really? You are still using Ubuntu 16.04? That advice actually worked more than 6 years ago...

Yes. When 16.04 was in business did you have access to sddm greeting with this method via X11VNC? It looks like you only had VNC access to whatever the environment loaded by SDDM. Once you exit this environment you should have seen sddm greeting and X11VNC crash, after that no VNC access to the box (to the greeting screen) until you again boot some environment. 
Normally once x11vnc is started you have a remote vnc access to X Server, including SDDM greeting box. So you can reboot the box then login with VNC to SDDM greeting box and proceed with any any environment you like. This is how x11vnc used to worked. 

The update above - fixes that problem. It persists over reboots, you can login to SDDM greeting box after reboot with VNC and proceed with the environment. 

SDDM developers are likely not familiar with x11vnc - they broke it. The vnc plugin to X is not perfect and lacks that functionality x11vnc has. Actually x11vnc is all we have now and SDDM broke it.

I didn't mean to hurt any feelings, sorry!

**Comment by meyergru** [comment]
My feelings were not hurt, for I did not give that advice. I only remember that it worked for me back then. 
I am quite aware that sddm broke x11vnc in the first place, but nowadays, this has long been fixed by two developments for me:

1. Subsequently, I switched to 18.04, 20.04 and now 22.04. I found solutions to make X11 work remotely for all of those.
2. I also moved to a virtualisation solution for all my servers (namely Proxmox), such that X11 is now within an HTML5 window inside my browser, accessible even from remote via a reverse proxy. That is far more convenient and stable than x11vnc ever was, let alone the advantage of snapshotting in case **any** update ever breaks anything again.

In other words: Life moves on...

**Comment by tenzap** [comment]
What about using `xauth extract - $DISPLAY ` ?

**Comment by ysalmon** [comment]
Inspired by @LanThrusteR 's [solution](https://github.com/sddm/sddm/issues/622#issuecomment-1301287521) but without init.d or a systemd service, I put this in ``/usr/share/sddm/scripts/Xsetup`` :

```
killall x11vnc
systemd-cat -t x11vnc-sddm /usr/bin/x11vnc -sleepin 0.5 -nopw -no6 -noxdamage -auth $(find /var/run/sddm/ -type f | head -n 1) -forever -shared -safer -display :0 &
```

and this seems to work well (I independently start a x11vnc instance when a user logs in, run as that user).
